### PR TITLE
Introduce 1967 proxy for upgrade

### PR DESCRIPTION
**Pull request by jac18281828**
#59 implementation of proxies in factory and builder

### CODE REVIEW

**Comment by jamalh110** [comment]
This is only for the factory right? Are we planning on adding proxies so that people can upgrade their deployed CG contracts?

**Comment by jac18281828** [comment]
> This is only for the factory right? Are we planning on adding proxies so that people can upgrade their deployed CG contracts?

Right, this introduces a 'standardized' means to upgrade the factories which are currently upgradable but not through the standardized mechanism.   Upgrade of the CG, or any other contract is part of a longer term discussion which seems to have some favorable opinions and some detractors.   Nevertheless the use of 1967 proxies in this manner gives us the ability to gain familiarity and confidence with these design patterns that might enable those future changes.

File: contracts/GovernanceFactory.sol [hunk 1]
```diff
@@ -97,7 +113,12 @@ contract GovernanceFactory is GovernanceFactoryCreator, UpgradeableContract, ERC
     function supportsInterface(bytes4 interfaceId) public view virtual override(IERC165, ERC165) returns (bool) {
         return
             interfaceId == type(GovernanceFactoryCreator).interfaceId ||
-            interfaceId == type(Upgradeable).interfaceId ||
+            interfaceId == type(Versioned).interfaceId ||
             super.supportsInterface(interfaceId);
     }
+
+    /// see UUPSUpgradeable
+    function _authorizeUpgrade(address _caller) internal virtual override(UUPSUpgradeable) onlyOwner {
```
Comment (rocketman-21) [inline contracts/GovernanceFactory.sol:121]: trying to understand the flow of this more eg: who the owner will likely be for a community treasury and what the process for upgrading will look like for the community

Comment (rocketman-21) [reply contracts/GovernanceFactory.sol:121]: i see the ownership class now, but if you could expand on this^ 🙏

Comment (jac18281828) [reply contracts/GovernanceFactory.sol:121]: There is some documentation for that here [Writing Upgradeable Contracts](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable) and here [UUPSUpgradeable](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d59306bd06a241083841c2e4a39db08e1f3722cc/contracts/proxy/utils/UUPSUpgradeable.sol#L94)

File: test/GovernanceFactory.t.sol [hunk 2]
```diff
@@ -119,8 +124,24 @@ contract GovernanceFactoryTest is Test {
...
     }
 
-    function testSupportsInterfaceUpgradeable() public {
-        bytes4 ifId = type(Upgradeable).interfaceId;
+    function testSupportsInterfaceVersioned() public {
+        bytes4 ifId = type(Versioned).interfaceId;
         assertTrue(_governanceFactory.supportsInterface(ifId));
     }
+
+    function testProxyUpgrade() public {
```
Comment (dannyadkins) [inline test/GovernanceFactory.t.sol:132]: Can someone explain what this test is doing?

**Review by dannyadkins — APPROVED** [review]

**Review by rocketman-21 — APPROVED** [review]
fyi i'm still getting up to speed on solidity, might miss some things. 

but looks good to me 👍

File: test/GovernanceFactory.t.sol (hunk 2 above)
Comment (jac18281828) [reply test/GovernanceFactory.t.sol:132]: 1. views the proxy as an upgradeable
2. upgrades the proxy to a new version that implements isUpgraded
3. calls the function confirming the upgrade worked

**Pull request merged**